### PR TITLE
fix #4163, remove `xdump` and simplify `dump`

### LINKED ISSUE #4163: tidy up dump and xdump
These functions have become a basket of odd behaviors, and their implementation is quite complex.

There are differences like this:

```
julia> dump(Int)
Int64

julia> xdump(Int)
Int64::DataType  <: Signed
```

Then there is this:

```
julia> dump(1+2im)
Complex{Int64} 
  re: Int64 1
  im: Int64 2

julia> dump((1+2im,{}))
(Complex{Int64},Array{Any,1}) (1 + 2im,{})
```

The first output was fine, but put it inside a tuple and something else happens.

Also `xdump` is not documented. If the current behavior does actually make sense, then the task here is to explain it by documenting `dump` and `xdump` better.

But I propose that we should only have `dump`, and it should simply traverse a structure and show everything about it.

### PR DESCRIPTION
**Pull request by JeffBezanson**
This replaces these functions with a single one that just shows everything inside an object.
Some examples:

```
julia> dump(("a",1+2im))
Tuple{ASCIIString,Complex{Int64}}
  1: ASCIIString
    data: Array(UInt8,(1,)) UInt8[0x61]
  2: Complex{Int64}
    re: Int64 1
    im: Int64 2

julia> dump(methods(sin).defs.func)
LambdaInfo
  code: Array(UInt8,(67,)) UInt8[0x10,0x88,0x88,0xcf,0x0e,0x10,0xb9,0x02,0x0a,0x66  …  0x1d,0x16,0x15,0x3b,0x0e,0x10,0x00,0x07,0x88,0x1d]
  slotnames: Array(Any,(2,))
    1: Symbol #self#
    2: Symbol a
  slottypes: Void nothing
  slotflags: Array(UInt8,(2,)) UInt8[0x00,0x00]
  gensymtypes: Int64 0
  rettype: Any
  sparam_syms: empty SimpleVector
  sparam_vals: empty SimpleVector
  tfunc: Void nothing
  name: Symbol sin
  roots: #undef
  specTypes: #undef
  unspecialized: #undef
  specializations: #undef
  module: Module Base
  def: LambdaInfo
# much longer but I'll cut it off here

julia> dump(StridedArray)
TypeConstructor
  parameters: SimpleVector
    1: TypeVar
      name: Symbol T
      lb: Union{}
      ub: Any
      bound: Bool false
    2: TypeVar
      name: Symbol N
      lb: Union{}
      ub: Any
      bound: Bool false
    3: TypeVar
      name: Symbol A
      lb: Union{}
      ub: DenseArray{T,N} <: AbstractArray{T,N}
      bound: Bool false
    4: TypeVar
      name: Symbol I
      lb: Union{}
      ub: Tuple{Vararg{Union{Base.AbstractCartesianIndex{N},Base.NoSlice,Colon,Int64,Range{Int64}}}} <: Any
      bound: Bool false
  body: Union{DenseArray{T,N},SubArray{T,N,A<:DenseArray{T,N},I<:Tuple{Vararg{Union{Base.AbstractCartesianIndex{N},Base.NoSlice,Colon,Int64,Range{Int64}}}},L}}
```

### CODE REVIEW

File: base/show.jl [hunk 1]
```diff
@@ -883,99 +883,115 @@ function show(io::IO, tv::TypeVar)
...
         if n > 0
-            for field in fieldnames(T)
-                if field != symbol("")    # prevents segfault if symbol is blank
-                    print(io, indent, "  ", field, ": ")
-                    if isdefined(x,field)
-                        fn(io, getfield(x, field), n - 1, string(indent, "  "))
-                    else
-                        println(io, undef_ref_str)
-                    end
+            for field in (isa(x,Tuple) ? (1:length(x)) : fieldnames(T))
```
Comment (vtjnash) [inline base/show.jl]: should probably always be `1:nfields(T)`, since the indices are unique and always available

**Comment by JeffBezanson** [comment]
One remaining issue is that `dump` was intended to be overloaded, and several packages do so. However I don't think this is a good idea; `dump` should show concrete structure, and all other printing can be handled by `show` at different verbosity levels.

**Comment by tshort** [comment]
So, `dump` becomes what `xdump` was. That seems like a loss of functionality for viewing contents of tree-like structures. Consider the following with the existing `dump` function:

```julia
julia> d = Dict(:a => 1, :b => Dict(:x => Dict(:y => [2,3], :z => 3)))
Dict{Symbol,Any} with 2 entries:
  :a => 1
  :b => Dict(:x=>Dict{Symbol,Any}(:y=>[2,3],:z=>3))

julia> dump(d)
Dict{Symbol,Any} len 2
  a: Int64 1
  b: Dict{Symbol,Dict{Symbol,Any}} len 1
    x: Dict{Symbol,Any} len 2
      y: Array(Int64,(2,)) [2,3]
      z: Int64 3

julia> show(d)
Dict{Symbol,Any}(:a=>1,:b=>Dict(:x=>Dict{Symbol,Any}(:y=>[2,3],:z=>3)))
```
If I follow this right, that output of `dump` will go away, and it'll show the internals of the dict. The current output is much more helpful than that printed by `show`. Can we have another function that does what `dump` does now?

**Comment by JeffBezanson** [comment]
That strikes me as just an aesthetic difference of opinion on how Dicts
should be printed. I think it could be handled by verbosity levels in show.
Actually the AbstractTrees package provides something like this, with much
nicer output to boot.

**Comment by ivirshup** [comment]
Can this be merged?

**Comment by ivirshup** [comment]
Is anything not mentioned here stopping this from being merged?

If not, I'd be up for branching off this and fixing it up for a merge.

**Comment by JeffBezanson** [comment]
Basically just needs a rebase and a NEWS entry. I'll do that and merge.

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -1692,7 +1692,7 @@ split
 """
     dump(x)
 
-Show all user-visible structure of a value.
+Show every part of the representation of a value.
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: please run genstdlib and commit updates at the same time

**Comment by tkelman** [comment]
several places in docs need fixing

**Comment by ivirshup** [comment]
What else needs docs past [`types.rst`](doc/devdocs/types.rst)?

**Comment by tkelman** [comment]
```
base/show.jl:1040:                #    xdump(fn, io, fieldtype(x,idx), n - 1, string(indent, "  "))
contrib/BBEditTextWrangler-julia.plist:1219:            <string>xdump</string>
doc/devdocs/types.rst:93:   julia> xdump(Array)
doc/devdocs/types.rst:107:   julia> xdump(T)
doc/devdocs/types.rst:136:   xdump(tv)
doc/devdocs/types.rst:172:   julia> xdump(p1[1].parameters[1])
doc/devdocs/types.rst:179:   julia> xdump(p3[1].parameters[1])
doc/devdocs/types.rst:313:   julia> xdump(Array.name)
doc/stdlib/io-network.rst:516:.. function:: xdump(x)
```

**Pull request merged**